File: LP5/C++/prac2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <omp.h>

using namespace std;

//g++ prac2.cpp -o prac2 -fopenmp

class ParallelSorting {
public:
    static void parallelBubbleSort(vector<int>& arr) {
        int n = arr.size();
        bool swapped;
        do {
            swapped = false;
            #pragma omp parallel for shared(arr, swapped)
            for (int i = 1; i < n; i++) {
                if (arr[i - 1] > arr[i]) {
                    swap(arr[i - 1], arr[i]);
                    swapped = true;
                }
            }
        } while (swapped);
    }

    static void parallelMergeSort(vector<int>& arr, int low, int high) {
        if (low < high) {
            int mid = (low + high) / 2;
            #pragma omp parallel sections
            {
                #pragma omp section
                parallelMergeSort(arr, low, mid);
                #pragma omp section
                parallelMergeSort(arr, mid + 1, high);
            }
            merge(arr, low, mid, high);
        }
    }

    static void merge(vector<int>& arr, int low, int mid, int high) {
        int n1 = mid - low + 1;
        int n2 = high - mid;

        vector<int> left(n1), right(n2);

        for (int i = 0; i < n1; i++) {
            left[i] = arr[low + i];
        }
        for (int j = 0; j < n2; j++) {
            right[j] = arr[mid + 1 + j];
        }

        int i = 0, j = 0, k = low;
        while (i < n1 && j < n2) {
            if (left[i] <= right[j]) {
                arr[k++] = left[i++];
            } else {
                arr[k++] = right[j++];
            }
        }

        while (i < n1) {
            arr[k++] = left[i++];
        }

        while (j < n2) {
            arr[k++] = right[j++];
        }
    }
};

int main() {
    const int size = 10000; // Change array size as needed
    vector<int> arr(size);
    generate(arr.begin(), arr.end(), []() { return rand(); });

    // Sequential Bubble Sort
    vector<int> arr1 = arr;
    double start = omp_get_wtime();
    sort(arr1.begin(), arr1.end());
    double end = omp_get_wtime();
    cout << "Sequential Bubble Sort: " << (end - start) * 1000 << " ms" << endl;

    // Parallel Bubble Sort
    vector<int> arr2 = arr;
    start = omp_get_wtime();
    ParallelSorting::parallelBubbleSort(arr2);
    end = omp_get_wtime();
    cout << "Parallel Bubble Sort: " << (end - start) * 1000 << " ms" << endl;

    // Sequential Merge Sort
    vector<int> arr3 = arr;
    start = omp_get_wtime();
    sort(arr3.begin(), arr3.end());
    end = omp_get_wtime();
    cout << "Sequential Merge Sort: " << (end - start) * 1000 << " ms" << endl;

    // Parallel Merge Sort
    vector<int> arr4 = arr;
    start = omp_get_wtime();
    ParallelSorting::parallelMergeSort(arr4, 0, arr4.size() - 1);
    end = omp_get_wtime();
    cout << "Parallel Merge Sort: " << (end - start) * 1000 << " ms" << endl;

    return 0;
}

File: LP5-Suraj/useless/prac2.py
import random
import time
import threading


class ParallelSorting:
    
    @staticmethod
    def bubble_sort(arr):
        n = len(arr)
        for i in range(n):
            for j in range(0, n-i-1):
                if arr[j] > arr[j+1]:
                    arr[j], arr[j+1] = arr[j+1], arr[j]
    
    @staticmethod
    def parallel_bubble_sort(arr):
        chunk_size = len(arr) // threading.active_count()
        chunks = [arr[i:i+chunk_size] for i in range(0, len(arr), chunk_size)]
        threads = []

        for chunk in chunks:
            thread = threading.Thread(target=ParallelSorting.bubble_sort, args=(chunk,))
            thread.start()
            threads.append(thread)

        for thread in threads:
            thread.join()

        # Merge sorted chunks
        sorted_arr = []
        for chunk in chunks:
            sorted_arr.extend(chunk)

        return sorted_arr

    @staticmethod
    def parallel_merge_sort(arr):
        if len(arr) > 1:
            mid = len(arr) // 2
            left = arr[:mid]
            right = arr[mid:]

            left_thread = threading.Thread(
                target=ParallelSorting.parallel_merge_sort, args=(left,))
            right_thread = threading.Thread(
                target=ParallelSorting.parallel_merge_sort, args=(right,))
            left_thread.start()
            right_thread.start()
            left_thread.join()
            right_thread.join()

            i = j = k = 0
            while i < len(left) and j < len(right):
                if left[i] < right[j]:
                    arr[k] = left[i]
                    i += 1
                else:
                    arr[k] = right[j]
                    j += 1
                k += 1

            while i < len(left):
                arr[k] = left[i]
                i += 1
                k += 1

            while j < len(right):
                arr[k] = right[j]
                j += 1
                k += 1


def main():
    size = 10000  # Change array size as needed
    arr = [random.randint(0, 1000) for _ in range(size)]

    # Sequential Bubble Sort
    arr1 = arr[:]
    start = time.time()
    arr1.sort()
    end = time.time()
    print("Sequential Bubble Sort:", (end - start) * 1000, "ms")

    # Parallel Bubble Sort
    arr2 = arr[:]
    start = time.time()
    ParallelSorting.parallel_bubble_sort(arr2)
    end = time.time()
    print("Parallel Bubble Sort:", (end - start) * 1000, "ms")

    # Sequential Merge Sort
    arr3 = arr[:]
    start = time.time()
    arr3.sort()
    end = time.time()
    print("Sequential Merge Sort:", (end - start) * 1000, "ms")

    # Parallel Merge Sort
    arr4 = arr[:]
    start = time.time()
    ParallelSorting.parallel_merge_sort(arr4)
    end = time.time()
    print("Parallel Merge Sort:", (end - start) * 1000, "ms")


if __name__ == "__main__":
    main()
